libraries/plugins/peerplays_sidechain/peerplays_sidechain_plugin.cpp

```diff
@@ -207,6 +211,113 @@ void peerplays_sidechain_plugin_impl::heartbeat_loop()
    schedule_heartbeat_loop();
 }
 
+chain::proposal_create_operation peerplays_sidechain_plugin_impl::create_son_down_proposal(chain::son_id_type son_id, fc::time_point_sec last_active_ts)
+{
+   chain::database& d = plugin.database();
+   chain::son_id_type my_son_id = *(_sons.begin());
+   const chain::global_property_object& gpo = d.get_global_properties();
+   const auto& idx = d.get_index_type<chain::son_index>().indices().get<by_id>();
+   auto son_obj = idx.find( my_son_id );
+
+   chain::son_report_down_operation son_down_op;
+   son_down_op.payer = gpo.parameters.get_son_btc_account_id();
+   son_down_op.son_id = son_id;
+   son_down_op.down_ts = last_active_ts;
+
+   proposal_create_operation proposal_op;
+   proposal_op.fee_paying_account = son_obj->son_account;
+   proposal_op.proposed_ops.push_back( op_wrapper( son_down_op ) );
+   uint32_t lifetime = ( gpo.parameters.block_interval * gpo.active_witnesses.size() ) * 3;
+   proposal_op.expiration_time = time_point_sec( d.head_block_time().sec_since_epoch() + lifetime );
+   return proposal_op;
+}
+
+void peerplays_sidechain_plugin_impl::on_block_applied( const signed_block& b )
+{
+   chain::database& d = plugin.database();
+   chain::son_id_type my_son_id = *(_sons.begin());
+   const chain::global_property_object& gpo = d.get_global_properties();
+   // Return if there are no active SONs
+   if(gpo.active_sons.size() <= 0) {
+      return;
+   }
+
+   chain::son_id_type next_son_id = d.get_scheduled_son(1);
+   if(next_son_id == my_son_id) {
+      const auto& idx = d.get_index_type<chain::son_index>().indices().get<by_id>();
+      for(auto son_id: gpo.active_sons) {
+         auto son_obj = idx.find( son_id );
+         auto stats = son_obj->statistics(d);
+         fc::time_point_sec last_active_ts = stats.last_active_timestamp;
+         int64_t down_threshold = 2*180000000;
+         if((fc::time_point::now() - last_active_ts) > fc::microseconds(down_threshold))  {
+            chain::proposal_create_operation op = create_son_down_proposal(son_id, last_active_ts);
+            chain::signed_transaction trx = d.create_signed_transaction(_private_keys.begin()->second, op);
+            fc::future<bool> fut = fc::async( [&](){
+               try {
+                  d.push_transaction(trx);
+                  plugin.app().p2p_node()->broadcast(net::trx_message(trx));
```

Comment:
Just asking, why do we need d.push_transaction and broadcast on p2p network?

Reply:
push_transaction would store the new transaction in the chain database _pending_txs which can be used when the next block is generated by us or validate tx when a block comes in from the network. 

broadcast would just post the transaction to the p2p network.